src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,215 @@
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.String)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                }
+
+                string key = reader.GetString()!;
 
-            if (metadata == MetadataPropertyName.Id)
+                // todo: verify value is T and throw JsonException? (currently no test)
+                value = (T)state.ReferenceResolver.ResolveReferenceOnDeserialize(key);
+
+                state.Current.ReturnValue = value;
+                state.Current.ObjectState = StackFrameObjectState.MetadataRefPropertyEndObject;
+            }
+            else if (state.Current.ObjectState == StackFrameObjectState.MetadataIdProperty)
             {
-                // Special case for dictionary properties since those do not push into the ReadStack.
-                // There is no need to check for enumerables since those will always be wrapped into JsonPreservableArrayReference<T> which turns enumerables into objects.
-                object value = state.Current.IsProcessingProperty(ClassType.Dictionary) ?
-                    state.Current.JsonPropertyInfo!.GetValueAsObject(state.Current.ReturnValue)! :
-                    state.Current.ReturnValue!;
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
 
-                state.ReferenceResolver.AddReferenceOnDeserialize(key, value);
+                if (reader.TokenType != JsonTokenType.String)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                }
+
+                state.Current.MetadataId = reader.GetString();
+
+                // Clear the MetadataPropertyName since we are done processing Id.
+                state.Current.JsonPropertyName = default;
+
+                if (converter.ClassType == ClassType.Enumerable)
+                {
+                    // Need to Read $values property name.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataValuesPropertyName;
+                }
+                else
+                {
+                    // We are done reading metadata.
+                    state.Current.ObjectState = StackFrameObjectState.MetadataPropertyValue;
+                    value = default!;
+                    return true;
+                }
             }
-            else if (metadata == MetadataPropertyName.Ref)
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefPropertyEndObject)
             {
-                state.Current.ReferenceId = key;
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.EndObject)
```

Comment:
Needless condition, after the read, the only possible `JsonTokenType`s are EndObject or `PropertyName` and you are already checking for the last.

Reply:
Yep thanks